### PR TITLE
Implemented analyzer to fix xml docs parameters

### LINKED ISSUE #357: Verify if xml docs have the correct parameters
As fixed on 526aa919a3afa369890ee9adfd81aae2216c5921.

Should work on methods and constructors.
Diagnostic should only be raised if there is already some xml documentation present.

### Removing unexistent param:

Before (`spans` does not exist):

````csharp
         /// </summary>
         /// <param name="analyzer">The analyzer to run on the documents</param>
         /// <param name="documents">The Documents that the analyzer will be run on</param>
         /// <param name="spans">Optional TextSpan indicating where a Diagnostic will be found</param>
         /// <returns>An IEnumerable of Diagnostics that surfaced in teh source code, sorted by Location</returns>
         protected async static Task<Diagnostic[]> GetSortedDiagnosticsFromDocumentsAsync(DiagnosticAnalyzer analyzer, Document[] documents)
         {
````

After:

````csharp
         /// </summary>
         /// <param name="analyzer">The analyzer to run on the documents</param>
         /// <param name="documents">The Documents that the analyzer will be run on</param>
         /// <returns>An IEnumerable of Diagnostics that surfaced in teh source code, sorted by Location</returns>
         protected async static Task<Diagnostic[]> GetSortedDiagnosticsFromDocumentsAsync(DiagnosticAnalyzer analyzer, Document[] documents)
         {
````

### Creating a missing parameter:

Before (note that `workspace` does not appear on the xml docs):

````csharp
         /// </summary>
         /// <param name="sources">Classes in the form of strings</param>
         /// <param name="language">The language the source code is in</param>
         /// <returns>A Project created out of the Douments created from the source strings</returns>
         public static Project CreateProject(string[] sources,
             out AdhocWorkspace workspace, string language = LanguageNames.CSharp)
````

After (with `workspace`):

````csharp
         /// </summary>
         /// <param name="sources">Classes in the form of strings</param>
         /// <param name="language">The language the source code is in</param>
         /// <param name="workspace">todo: describe workspace parameter on CreateProject</param>
         /// <returns>A Project created out of the Douments created from the source strings</returns>
         public static Project CreateProject(string[] sources,
             out AdhocWorkspace workspace, string language = LanguageNames.CSharp)
````

When including a parameter leave the description in the format:

````
todo: describe {parameter} parameter on {method}
````

The code actions listed should be:

 * One for each missing parameter
 * One for all parameters that are missing

Severity: `Warning`
Diagnostic id: `CC0090`
Category: `Maintenability`

### PR DESCRIPTION
**Pull request by AlbertoMonteiro**
I created analyzer and 2 code fixes, one for missing parameters and another to unexisting parameters.

Tests are ok, but I am facing a problem when testing running VS 

But when I press <kbd>CTRL</kbd>+<kbd>.</kbd> I get a error in VS.
**"One or more erros ocurred."**

When I checked to see all exceptions in VS I got StackTrace and got this exception:
````
System.InvalidCastException occurred
Message: Exception thrown: 'System.InvalidCastException' in mscorlib.dll
Additional information: Unable to cast object of type 'System.Reflection.RuntimeMethodInfo' to type 'System.Reflection.ConstructorInfo'.
````

I tried a lot things, but couldnt fix, I need help.
Fix #357

### CODE REVIEW

File: src/CSharp/CodeCracker/Refactoring/XmlDocumentationAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        internal const string Category = SupportedCategories.Refactoring;
+
+        internal static DiagnosticDescriptor XmlDocumentationRule = new DiagnosticDescriptor(
+            DiagnosticId.XmlDocumentation.ToDiagnosticId(),
+            Title,
+            MessageFormat,
+            SupportedCategories.Refactoring,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            helpLinkUri: HelpLink.ForDiagnostic(DiagnosticId.XmlDocumentation));
```
Comment (giggio) [inline src/CSharp/CodeCracker/Refactoring/XmlDocumentationAnalyzer.cs]: Category should be maintenability. Needs to change the `DiagnosticSeverity` and file locations.

File: src/CSharp/CodeCracker/Refactoring/XmlDocumentationCreateMissingParametersCodeFixProvider.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+        private static SyntaxList<XmlAttributeSyntax> CreateXmlAttributes(string paramenterName)
+        {
+            var syntaxToken = SyntaxFactory.Token(DoubleQuoteToken);
+            var xmlAttributeName = SyntaxFactory.XmlName(SyntaxFactory.Identifier(@"name"));
+
+            return SyntaxFactory.SingletonList<XmlAttributeSyntax>(SyntaxFactory.XmlNameAttribute(xmlAttributeName, syntaxToken, SyntaxFactory.IdentifierName(paramenterName), syntaxToken));
+        }
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
```
Comment (giggio) [inline src/CSharp/CodeCracker/Refactoring/XmlDocumentationCreateMissingParametersCodeFixProvider.cs]: `RegisterCodeFixesAsync` should be sealed.

File: src/CSharp/CodeCracker/Refactoring/XmlDocumentationRemoveUnexistentParametersCodeFixProvider.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            var allNodesToRemove = nodesToRemove.Cast<SyntaxNode>().Union(xmlTextNodesToRemove);
+            return allNodesToRemove;
+        }
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var diagnostic = context.Diagnostics.First();
+            if (diagnostic.Properties["kind"] == "unexistent")
+                context.RegisterCodeFix(CodeAction.Create(MessageFormat, c => FixParametersAsync(context.Document, diagnostic, c)), diagnostic);
+            return Task.FromResult(0);
```
Comment (giggio) [inline src/CSharp/CodeCracker/Refactoring/XmlDocumentationRemoveUnexistentParametersCodeFixProvider.cs]: `RegisterCodeFixesAsync` should be sealed.

File: src/CSharp/CodeCracker/Refactoring/XmlDocumentationCodeFixProvider.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+namespace CodeCracker.CSharp.Refactoring
+{
+    public abstract class XmlDocumentationCodeFixProvider : CodeFixProvider
+    {
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(DiagnosticId.XmlDocumentation.ToDiagnosticId());
+
+        public abstract SyntaxNode FixParameters(MethodDeclarationSyntax method, SyntaxNode root);
+
+        public abstract string MessageFormat { get; }
```
Comment (giggio) [inline src/CSharp/CodeCracker/Refactoring/XmlDocumentationCodeFixProvider.cs]: No need for this property to exist on the base class, it is not used here. It can be a const string field (not a property) on the derived classes

File: src/CSharp/CodeCracker/Refactoring/XmlDocumentationCreateMissingParametersCodeFixProvider.cs (hunk 2 above)
Comment (sharwell) [inline src/CSharp/CodeCracker/Refactoring/XmlDocumentationCreateMissingParametersCodeFixProvider.cs]: :question: Why?

File: src/CSharp/CodeCracker/Refactoring/XmlDocumentationCodeFixProvider.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            var methodParameters = method.ParameterList.Parameters;
+
+            var xElementsWitAttrs = documentationNode.Content.OfType<XmlElementSyntax>()
+                                    .Where(xEle => xEle.StartTag.Name.LocalName.ValueText == "param")
+                                    .SelectMany(xEle => xEle.StartTag.Attributes, (xEle, attr) => new Tuple<XmlElementSyntax, XmlNameAttributeSyntax>(xEle, (XmlNameAttributeSyntax)attr));
+
+            var methodParameterLookUp = methodParameters.ToLookup(parameter => parameter.Identifier.ValueText);
+            var docParameterLookUp = xElementsWitAttrs.ToLookup(x => x.Item2.Identifier.Identifier.ValueText);
+
+            var keys = methodParameterLookUp.Select(p => p.Key).Union(docParameterLookUp.Select(p => p.Key)).ToImmutableHashSet();
```
Comment (giggio) [inline src/CSharp/CodeCracker/Refactoring/XmlDocumentationCodeFixProvider.cs]: It does not seem necessary to create 2 lookups if the only reason for their existence is to later create a union on their keys. Analyzers are all about performance, simply create a hash from the parameter identifiers and xelements without creating the intermediate lookup.

File: src/CSharp/CodeCracker/Refactoring/XmlDocumentationAnalyzer.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            var method = documentationNode.AncestorsAndSelf().OfType<MethodDeclarationSyntax>().First();
+            var methodParameters = method.ParameterList.Parameters;
+            var xElementsWitAttrs = documentationNode.Content.OfType<XmlElementSyntax>()
+                                    .Where(xEle => xEle.StartTag.Name.LocalName.ValueText == "param")
+                                    .SelectMany(xEle => xEle.StartTag.Attributes, (xEle, attr) => (XmlNameAttributeSyntax)attr);
+
+            var methodParameterLookUp = methodParameters.ToLookup(parameter => parameter.Identifier.ValueText);
+            var docParameterLookUp = xElementsWitAttrs.ToLookup(attr => attr.Identifier.Identifier.ValueText);
+
+            var keys = methodParameterLookUp.Select(p => p.Key).Union(docParameterLookUp.Select(p => p.Key)).ToImmutableHashSet();
```
Comment (giggio) [inline src/CSharp/CodeCracker/Refactoring/XmlDocumentationAnalyzer.cs]: It does not seem necessary to create 2 lookups if the only reason for their existence is to later create a union on their keys. Analyzers are all about performance, simply create a hash from the parameter identifiers and xelements without creating the intermediate lookup.

File: src/CSharp/CodeCracker/Refactoring/XmlDocumentationAnalyzer.cs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+            {
+                var properties = new Dictionary<string, string> {["kind"] = "unexistent" }.ToImmutableDictionary();
+                var diagnostic = Diagnostic.Create(XmlDocumentationRule, documentationNode.GetLocation(), properties, properties["kind"]);
+                context.ReportDiagnostic(diagnostic);
+            }
+
+            if (paramterWithDocParameter.Any(p => p.DocParameter == null))
+            {
+                var properties = new Dictionary<string, string> { ["kind"] = "missing" }.ToImmutableDictionary();
+                var diagnostic = Diagnostic.Create(XmlDocumentationRule, documentationNode.GetLocation(), properties, properties["kind"]);
```
Comment (giggio) [inline src/CSharp/CodeCracker/Refactoring/XmlDocumentationAnalyzer.cs]: No need for the last parameter on the `Diagnostic.Create` invocation, you are not using it on the `MessageFormat`.

Comment (giggio) [inline src/CSharp/CodeCracker/Refactoring/XmlDocumentationAnalyzer.cs]: I am finding it difficult to follow the `missing` and `unexistent` texts. They almost mean the same thing. Can you try to be more descriptive? Maybe `missingDoc` and `nonexistentParam`, or something like that?

File: src/CSharp/CodeCracker/Refactoring/XmlDocumentationRemoveUnexistentParametersCodeFixProvider.cs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace CodeCracker.CSharp.Refactoring
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, nameof(XmlDocumentationCodeFixProvider)), Shared]
+    public class XmlDocumentationRemoveUnexistentParametersCodeFixProvider : XmlDocumentationCodeFixProvider
```
Comment (giggio) [inline src/CSharp/CodeCracker/Refactoring/XmlDocumentationRemoveUnexistentParametersCodeFixProvider.cs]: Use `NonExistent` instead of `Unexistent`.

File: src/CSharp/CodeCracker/Refactoring/XmlDocumentationCodeFixProvider.cs [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        protected static IEnumerable<Tuple<ParameterSyntax, Tuple<XmlElementSyntax, XmlNameAttributeSyntax>>> GetMethodParametersWithDocParameters(MethodDeclarationSyntax method, DocumentationCommentTriviaSyntax documentationNode)
+        {
+            var methodParameters = method.ParameterList.Parameters;
+
+            var xElementsWitAttrs = documentationNode.Content.OfType<XmlElementSyntax>()
+                                    .Where(xEle => xEle.StartTag.Name.LocalName.ValueText == "param")
+                                    .SelectMany(xEle => xEle.StartTag.Attributes, (xEle, attr) => new Tuple<XmlElementSyntax, XmlNameAttributeSyntax>(xEle, (XmlNameAttributeSyntax)attr));
+
+            var methodParameterLookUp = methodParameters.ToLookup(parameter => parameter.Identifier.ValueText);
+            var docParameterLookUp = xElementsWitAttrs.ToLookup(x => x.Item2.Identifier.Identifier.ValueText);
```
Comment (giggio) [inline src/CSharp/CodeCracker/Refactoring/XmlDocumentationCodeFixProvider.cs]: You don't need the Tuple if all you are doing is accessing `Item2`. Simply return it from the `SelectMany` above.

File: src/CSharp/CodeCracker/Refactoring/XmlDocumentationCodeFixProvider.cs [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(DiagnosticId.XmlDocumentation.ToDiagnosticId());
+
+        public abstract SyntaxNode FixParameters(MethodDeclarationSyntax method, SyntaxNode root);
+
+        public abstract string MessageFormat { get; }
+
+        protected async Task<Document> FixParametersAsync(Document document, Diagnostic diagnostic, CancellationToken c)
+        {
+            var root = await document.GetSyntaxRootAsync(c).ConfigureAwait(false);
+            var documentationNode = (MethodDeclarationSyntax)root.FindNode(diagnostic.Location.SourceSpan);
```
Comment (giggio) [inline src/CSharp/CodeCracker/Refactoring/XmlDocumentationCodeFixProvider.cs]: I caught an exception on this line:

````cshap
var documentationNode = (MethodDeclarationSyntax)root.FindNode(diagnostic.Location.SourceSpan);
````

Details:

````
Exception thrown: 'System.InvalidCastException' in CodeCracker.CSharp.dll

Additional information: Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.Syntax.PredefinedTypeSyntax' to type 'Microsoft.CodeAnalysis.CSharp.Syntax.MethodDeclarationSyntax'.
````

The problem is the node being found is the return type of the method. Even though the return type has a method **parent**, it is not the `MethodDeclaration` that is being found by the `FindNode` method. The most specific type is always found. Try to use `root.FindToken(diagnostic.Location.SourceSpan.Start).Parent.AncestorsAndSelf().OfType<MethodDeclarationSyntax>().First()` instead.

File: src/CSharp/CodeCracker/Refactoring/XmlDocumentationCreateMissingParametersCodeFixProvider.cs (hunk 2 above)
Comment (giggio) [inline src/CSharp/CodeCracker/Refactoring/XmlDocumentationCreateMissingParametersCodeFixProvider.cs]: @sharwell 3 reasons: style, recommendation and performance.
Style: All our `RegisterCodeFixesAsync` are sealed.
Recommendation: this is how Microsoft is recommending on their own template, and this is how it is done on Roslyn itself.
Performance: while the compiler usually emits IL with a `callvirt` to sealed and non-virtual method calls, the method is still marked `final`, which is a hint for the JIT later for possible optimizations, like method inlining.

File: src/CSharp/CodeCracker/Refactoring/XmlDocumentationAnalyzer.cs (hunk 1 above)
Comment (AlbertoMonteiro) [inline src/CSharp/CodeCracker/Refactoring/XmlDocumentationAnalyzer.cs]: Which  ````DiagnosticSeverity````  should I set?

File: src/CSharp/CodeCracker/Refactoring/XmlDocumentationCodeFixProvider.cs (hunk 9 above)
Comment (AlbertoMonteiro) [inline src/CSharp/CodeCracker/Refactoring/XmlDocumentationCodeFixProvider.cs]: I do it in method GetAllNodesToRemove on ````XmlDocumentationRemoveNonExistentParametersCodeFixProvider````

````csharp
var nodesToRemove = paramterWithDocParameter.Where(p => p.Item1 == null).Select(x => x.Item2.Item1).ToList();
````

**Comment by AlbertoMonteiro** [comment]
@giggio I made many commits to fix your observations. When everything is aproved I will squash them all.

**Comment by AlbertoMonteiro** [comment]
@giggio How did you catch this exception in comment: https://github.com/code-cracker/code-cracker/pull/403#discussion_r33905132 ?
After these commits to fix your observations, I ran VS and it fixed the <kbd>CTRL</kbd> + <kbd>.</kbd>

I walk through all my new commits and discovered that <kbd>CTRL</kbd> + <kbd>.</kbd> bug stoped in https://github.com/AlbertoMonteiro/code-cracker/commit/ebdae16c005a967b782bd921c514466a21072721

So I started to mix the changes that I did in this commit, and I discovered that if I use the namespace **CodeCracker.CSharp.Refactoring** it will break when running VS.
So I moved to parent commit(https://github.com/AlbertoMonteiro/code-cracker/commit/150a201a5a21f39b32185b63e049261f3a8b1357) and ran VS again, the bug happend, stoped, and change the namespace from **Refactoring** to **AnyElse** and the problem was gone.

Why the simple namespace name change, breack when running on VS?

**Comment by giggio** [comment]
@AlbertoMonteiro You set the exception settings to catch all .NET Exceptions. If you don't do that, VS will not stop to let you inspect the exception. I don't know why, it should stop.
BTW, you also need to update the analyzer and code fix to use Resources for globalization support. See https://github.com/code-cracker/code-cracker/blob/cc44ad2419e25d794471b83867fbd08877af25dd/src/CSharp/CodeCracker/Style/ConsoleWriteLineAnalyzer.cs and the corresponding code fix.

**Comment by giggio** [comment]
@AlbertoMonteiro I can't see why the namespace would break the code. Try using the exception setting I talked about above and see if you can find what happened. Anyway, you are going to move it another namespace, so maybe this is moot.

File: src/CSharp/CodeCracker/Refactoring/XmlDocumentationCreateMissingParametersCodeFixProvider.cs (hunk 2 above)
Comment (sharwell) [inline src/CSharp/CodeCracker/Refactoring/XmlDocumentationCreateMissingParametersCodeFixProvider.cs]: :question: Why not just make the class sealed?

**Comment by AlbertoMonteiro** [comment]
@giggio commited the globalization support.
About the namespace breaking, I will record a vídeo and upload on youtube , so you can see this weird behavior.

File: src/CSharp/CodeCracker/Refactoring/XmlDocumentationCodeFixProvider.cs (hunk 9 above)
Comment (giggio) [inline src/CSharp/CodeCracker/Refactoring/XmlDocumentationCodeFixProvider.cs]: Ok, got it.

**Comment by giggio** [comment]
@AlbertoMonteiro Good job!
I did some small updates, check e1d2286 for more info. I got some edge cases that needed taking care of and the test file was on the wrong directory.

**Comment by AlbertoMonteiro** [comment]
@giggio Thanks.
I will pay more antention next time!
